From zuoqian113@gmail.com  Thu Jul 30 15:29:33 2026
From: Qian Zuo <zuoqian113@gmail.com>
Date: Tue,  7 Jul 2026 03:46:08 +0000
Subject: coredump: fix pidfs file refcount leak in umh_coredump_setup
To: viro@zeniv.linux.org.uk, brauner@kernel.org
Cc: gregkh@linuxfoundation.org, oleg@redhat.com, linux-fsdevel@vger.kernel.org, linux-kernel@vger.kernel.org, stable@vger.kernel.org, Qian Zuo <zuoqian113@gmail.com>
Message-ID: <20260707034608.912-1-zuoqian113@gmail.com>

From: Qian Zuo <zuoqian113@gmail.com>

The backport of upstream commit b5325b2a270f introduced a reference
count leak for pidfs_file.

In the upstream implementation, pidfs_file is declared with
__free(fput), which automatically drops the initial file reference when
leaving the scope. During the backport, the code was rewritten to manage
the file pointer manually, but after a successful replace_fd(),
pidfs_file was simply cleared without dropping the initial reference.

As a result, the pidfs file keeps an extra reference and is never
released after the usermode helper exits, leaving associated objects,
such as struct pid, permanently allocated and triggering kmemleak
reports.

Fix this by explicitly calling fput(pidfs_file) after replace_fd().

Fixes: cdb61a705f5f ("coredump: hand a pidfd to the usermode coredump helper")
Signed-off-by: Qian Zuo <zuoqian113@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 fs/coredump.c |    1 +
 1 file changed, 1 insertion(+)

--- a/fs/coredump.c
+++ b/fs/coredump.c
@@ -560,6 +560,7 @@ static int umh_coredump_setup(struct sub
 		if (err < 0)
 			goto out_fail;
 
+		fput(pidfs_file);
 		pidfs_file = NULL;
 	}
 
